### PR TITLE
Add debug logging in ClassfileManager

### PR DESCRIPTION
**Pull request by gkossakowski**
Add logging of various operations the transactional class file manager is
doing. You can pass logger to be used by the transactional class file
manager by using overloaded definition of `ClassfileManager.transactional`
method. The old overload has been deprecated.

The factory methods for class file manager in IncOptions companion object
has been deprecated in favor of using ClassfileManager companion object
directly. The code in Defaults.scala has been updated to use non-deprecated
methods.

The canonical way of enabling transactional class file manager in sbt
project is:

``` scala
incOptions := incOptions.value.withNewClassfileManager(
  sbt.inc.ClassfileManager.transactional(
    crossTarget.value / "classes.bak",
    (streams in (Compile, compile)).value.log
  )
)
```

It's a bit verbose which shows that the api for this is not the best.
However, I don't expect sbt users to need this code very often.

This patch should help debug the problem described in #1184

### CODE REVIEW

**Comment by gkossakowski** [comment]
Review by @jsuereth

File: main/src/main/scala/sbt/Defaults.scala [hunk 1]
```diff
@@ -215,7 +215,8 @@ object Defaults extends BuildCommon
 	)
 
 	def compileBase = inTask(console)(compilersSetting :: Nil) ++ compileBaseGlobal ++ Seq(
-		incOptions := IncOptions.setTransactional(incOptions.value, crossTarget.value / "classes.bak"),
+		incOptions := incOptions.value.withNewClassfileManager(
+			sbt.inc.ClassfileManager.transactional(crossTarget.value / "classes.bak", sbt.Logger.Null)),
```
Comment (jsuereth) [inline main/src/main/scala/sbt/Defaults.scala:219]: should this one be `streams.value.log` instead of `sbt.Logger.Null` ?

**Comment by jsuereth** [comment]
your example may want to show:

``` scala
(streams in (compile, Compile)).value.log
```

Or some such variant if you want the log to align with the `compile in Compile` task.   The stream-logging stuff is a bit... fun right now, and needs some cleanup.  But just curious if the output locations is:

a. where you want it
b. in the same file as the `compile in Compile` logs.

**Comment by jsuereth** [comment]
Oh, otherwise LGTM

File: main/src/main/scala/sbt/Defaults.scala (hunk 1 above)
Comment (gkossakowski) [inline main/src/main/scala/sbt/Defaults.scala:219]: Logging is opt-in. I didn't want to clutter the debug log too much. Most of the time you don't really need to know what ClassfileManager does even if you are debugging incremental compiler problems.

**Comment by gkossakowski** [comment]
Clarified in commit message that logging is off by default and fixed the code snippet according to your suggestion.

I agree that logging in sbt could receive some love.

**Comment by aloiscochard** [comment]
@gkossakowski the instruction you provided did not work.

I had to do:

```
    incOptions := incOptions.value.withNewClassfileManager(
      sbt.inc.ClassfileManager.transactional(
        crossTarget.value / "classes.bak",
        (streams in Compile).value.log
      )
    )
```

I prefer not to know why it should be like that, but I thought it was worth mentioning if anyone else try to use this.

**Comment by gkossakowski** [comment]
@jsuereth: any ideas on why the original snippet would not work?

**Comment by jsuereth** [comment]
@gkossakowski I sent you an email about it elsewhere.   You have the wrong order for `in` so you confuse the hell out of scala looking for implicits on the overloaded method.

Your code _should_ have been:

```
incOptions := incOptions.value.withNewClassfileManager(
  sbt.inc.ClassfileManager.transactional(
    crossTarget.value / "classes.bak",
    (streams in (Compile, compile)).value.log
  )
)
```

See me note above.

**Comment by gkossakowski** [comment]
@jsuereth: I updated the description in this issue. Thanks.

**Pull request merged**